pkg/internal/controller/controller.go

```diff
@@ -99,17 +100,21 @@ type watchDescription struct {
 
 // Reconcile implements reconcile.Reconciler.
 func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, err error) {
-	if c.RecoverPanic {
-		defer func() {
-			if r := recover(); r != nil {
+	log := c.Log.WithValues("name", req.Name, "namespace", req.Namespace)
+	defer func() {
+		if r := recover(); r != nil {
+			if c.RecoverPanic {
 				for _, fn := range utilruntime.PanicHandlers {
 					fn(r)
 				}
 				err = fmt.Errorf("panic: %v [recovered]", r)
+				return
 			}
-		}()
-	}
-	log := c.Log.WithValues("name", req.Name, "namespace", req.Namespace)
+
+			log.Info(fmt.Sprintf("Observed a panic: %v", r))
```

Comment:
This line only prints the panic message like `Invalid memory address or nil pointer dereference`. Should we also use `utilruntime.PanicHandlers` to print the panic stack, so that user can track where it happened?

Reply:
I'm not sure. The panic in l.115 will print the full panic stack, though.

I don't know enough about the idea behind PanicHandlers if it should be called it that case too.

Previously the "API" was that they are only called if RecoveryPanic was set to true.

Reply:
> The panic in l.115 will print the full panic stack

But if the panic itself is enough, why should we recover and panic it again? This just add the duplicated message print.

Reply:
The benefit is that this log line correlates the panic with a concrete reconciled object.

For example we had the situation that a controller panic'ed at some point, but as there were ~50 objects which could have been reconciled at that moment it was impossible to figure out which object triggered the panic without a go debugger. The logger at this time contains the namespace and name of the current object, which makes it trivial to see which object "triggers" the panic.